File: Chapter 11/JS/section_11_4_secrets_management/secure/application/orderService.js
/**
 * APPLICATION LAYER - Order Service
 * 
 * This is the application service that orchestrates business logic.
 * It depends on the SecretsPort abstraction (Port), not on any
 * specific secrets manager implementation.
 * 
 * ARCHITECTURAL NOTE:
 * This is the Application Layer in Clean Architecture. It:
 * 1. Contains business logic
 * 2. Depends on abstractions (SecretsPort), not concretions
 * 3. Doesn't know HOW secrets are retrieved, only THAT they can be retrieved
 * 
 * SECURITY BENEFITS:
 * - No credentials in source code
 * - Credentials rotated automatically by AWS
 * - Least privilege IAM role
 * - Full audit trail in CloudTrail
 */

import { SecretNotFoundError, PermissionError } from '../domain/secretsPort.js';

export class OrderService {
    /**
     * Initialize the Order Service with a secrets provider.
     * 
     * @param {SecretsPort} secretsProvider - Implementation of SecretsPort (injected via DI)
     */
    constructor(secretsProvider) {
        this.secretsProvider = secretsProvider;
    }

    /**
     * Connect to the database using credentials from the secrets manager.
     * 
     * @param {string} secretName - The name of the secret in AWS Secrets Manager
     * @returns {Promise<{success: boolean, message: string}>} Result of connection attempt
     */
    async connectToDatabase(secretName) {
        try {
            // Retrieve credentials from the secrets manager
            // The application code only contains the SECRET NAME, not the secret itself!
            console.log("  [OrderService] Requesting credentials from secrets manager...");
            console.log(`  [OrderService] Secret name: ${secretName}`);
            
            const { username, password } = await this.secretsProvider.getDatabaseCredentials(secretName);
            
            console.log("  [OrderService] Credentials retrieved successfully");
            console.log(`  [OrderService] Connecting to database with username: ${username}`);
            
            // In a real application, we would now use these credentials to connect
            // For demo purposes, we'll simulate a successful connection
            // const { Client } = await import('pg');
            // const client = new Client({
            //     host: 'prod-db.internal.castle.com',
            //     database: 'orders_db',
            //     user: username,
            //     password: password
            // });
            
            return { success: true, message: `Successfully connected to database as ${username}` };
            
        } catch (error) {
            if (error instanceof SecretNotFoundError) {
                const errorMsg = `Secret not found: ${error.message}`;
                console.log(`  [OrderService] ERROR: ${errorMsg}`);
                return { success: false, message: errorMsg };
            } else if (error instanceof PermissionError) {
                const errorMsg = `Permission denied: ${error.message}`;
                console.log(`  [OrderService] ERROR: ${errorMsg}`);
                return { success: false, message: errorMsg };
            } else {
                const errorMsg = `Unexpected error: ${error.message}`;
                console.log(`  [OrderService] ERROR: ${errorMsg}`);
                return { success: false, message: errorMsg };
            }
        }
    }

    /**
     * Process an order by connecting to the database and updating the order.
     * 
     * @param {string} orderId - The order to process
     * @param {string} secretName - The secret name for database credentials
     * @returns {Promise<string>} Status message
     */
    async processOrder(orderId, secretName) {
        console.log(`\n  [OrderService] Processing order ${orderId}...`);
        
        // Connect to database using secure credentials
        const { success, message } = await this.connectToDatabase(secretName);
        
        if (!success) {
            return `Failed to process order ${orderId}: ${message}`;
        }
        
        // Simulate order processing
        console.log("  [OrderService] Updating order status to 'PROCESSED'...");
        console.log(`  [OrderService] Order ${orderId} processed successfully!`);
        
        return `Order ${orderId} processed successfully`;
    }
}

File: Chapter 11/Python/section_11_4_secrets_management/secure/application/order_service.py
"""
APPLICATION LAYER - Order Service

This is the application service that orchestrates business logic.
It depends on the ISecretsProvider abstraction (Port), not on any
specific secrets manager implementation.
"""

from typing import Tuple
from ..domain.secrets_port import ISecretsProvider, SecretNotFoundError, SecretPermissionError


class OrderService:
    """
    APPLICATION SERVICE: Orchestrates order-related operations.
    
    ARCHITECTURAL NOTE:
    This is the Application Layer in Clean Architecture. It:
    1. Contains business logic
    2. Depends on abstractions (ISecretsProvider), not concretions
    3. Doesn't know HOW secrets are retrieved, only THAT they can be retrieved
    
    SECURITY BENEFITS:
    - No credentials in source code
    - Credentials rotated automatically by AWS
    - Least privilege IAM role
    - Full audit trail in CloudTrail
    """
    
    def __init__(self, secrets_provider: ISecretsProvider):
        """
        Initialize the Order Service with a secrets provider.
        
        Args:
            secrets_provider: Implementation of ISecretsProvider (injected via DI)
        """
        self.secrets_provider = secrets_provider
    
    def connect_to_database(self, secret_name: str) -> Tuple[bool, str]:
        """
        Connect to the database using credentials from the secrets manager.
        
        Args:
            secret_name: The name of the secret in AWS Secrets Manager
            
        Returns:
            Tuple of (success: bool, message: str)
        """
        try:
            # Retrieve credentials from the secrets manager
            # The application code only contains the SECRET NAME, not the secret itself!
            print(f"  [OrderService] Requesting credentials from secrets manager...")
            print(f"  [OrderService] Secret name: {secret_name}")
            
            username, password = self.secrets_provider.get_database_credentials(secret_name)
            
            print(f"  [OrderService] Credentials retrieved successfully")
            print(f"  [OrderService] Connecting to database with username: {username}")
            
            # In a real application, we would now use these credentials to connect
            # For demo purposes, we'll simulate a successful connection
            # conn = psycopg2.connect(
            #     host="prod-db.internal.castle.com",
            #     database="orders_db",
            #     user=username,
            #     password=password
            # )
            
            return True, f"Successfully connected to database as {username}"
            
        except SecretNotFoundError as e:
            error_msg = f"Secret not found: {e}"
            print(f"  [OrderService] ERROR: {error_msg}")
            return False, error_msg
            
        except SecretPermissionError as e:
            error_msg = f"Permission denied: {e}"
            print(f"  [OrderService] ERROR: {error_msg}")
            return False, error_msg
            
        except Exception as e:
            error_msg = f"Unexpected error: {e}"
            print(f"  [OrderService] ERROR: {error_msg}")
            return False, error_msg
    
    def process_order(self, order_id: str, secret_name: str) -> str:
        """
        Process an order by connecting to the database and updating the order.
        
        Args:
            order_id: The order to process
            secret_name: The secret name for database credentials
            
        Returns:
            Status message
        """
        print(f"\n  [OrderService] Processing order {order_id}...")
        
        # Connect to database using secure credentials
        success, message = self.connect_to_database(secret_name)
        
        if not success:
            return f"Failed to process order {order_id}: {message}"
        
        # Simulate order processing
        print(f"  [OrderService] Updating order status to 'PROCESSED'...")
        print(f"  [OrderService] Order {order_id} processed successfully!")
        
        return f"Order {order_id} processed successfully"